### PR TITLE
Use hdfsPreadFully to reduce the amount of JNI array allocation for object storage

### LINKED ISSUE #1336: The performance of Scan S3/OSS is slow
We test the scan performance of S3/OSS using TPC-H.
The scale is 100. And the data is stored as the ORC format using snappy compression.

Environment:
   AWS m5.4xlarge (16vcpu, 8core, 64G)

### PR DESCRIPTION
**Pull request by miomiocat**
libhdfs in version 3.3.0 contains 3 interfaces to read data: hdfsRead(), hdfsPread(),  ​hdfsPreadFully().
   1. hdfsRead() is the default read interface to read HDFS. 
   2. hdfsPread() is used to HDFS hedged reads. About HDFS hedged reads: https://issues.apache.org/jira/browse/HDFS-5776
   3. hdfsPreadFully() will "Read the specified number of bytes" instead of hdfsPread() will "Read up to the specified number of bytes".

If the underlying FileSystem does not support ByteBuffer reads (HDFS-2834) (e.g., S3A/OSS does not support this feature), then hdfsPread() will allocate a Java array equal in size to the specified length of the buffer; the call to hdfsPread() may only fill up the buffer partially. Under this circumstance, it will repeat the call to hdfsPread() since the buffer was not filled, which will cause another large array allocation; this can result in a lot of wasted time doing unnecessary array allocations. The idea comes from https://issues.apache.org/jira/browse/IMPALA-8525.

hdfsPreadFully() will be default option to read HDFS. The configure `user_hdfs_pread` can be used to enable hdfsRead() instead of  hdfsPreadFully().

### CODE REVIEW

File: be/src/env/env_hdfs.cpp [hunk 1]
```diff
@@ -50,19 +51,28 @@ static Status read_at_internal(hdfsFS fs, hdfsFile file, const std::string& file
                                                        offset, get_hdfs_err_msg()));
         }
     }
-    size_t bytes_read = 0;
-    while (bytes_read < res->size) {
-        size_t to_read = res->size - bytes_read;
-        auto hdfs_res = hdfsRead(fs, file, res->data + bytes_read, to_read);
-        if (hdfs_res < 0) {
+    if (starrocks::config::use_hdfs_pread || is_object_storage_path(file_name.c_str())) {
```
Comment (imay) [inline be/src/env/env_hdfs.cpp:54]: This will cause string compare for each read. Can you eliminate this compare?

Comment (miomiocat) [reply be/src/env/env_hdfs.cpp]: done, PTAL

File: be/src/env/env_hdfs.cpp [hunk 2]
```diff
@@ -38,44 +38,52 @@ void HdfsRandomAccessFile::close() noexcept {
...
-    while (bytes_read < res->size) {
-        size_t to_read = res->size - bytes_read;
-        auto hdfs_res = hdfsRead(fs, file, res->data + bytes_read, to_read);
-        if (hdfs_res < 0) {
-            return Status::IOError(
-                    strings::Substitute("fail to read file, file=$0, error=$1", file_name, get_hdfs_err_msg()));
-        } else if (hdfs_res == 0) {
-            break;
+    if (usePread) {
+        if (hdfsPreadFully(fs, file, offset, res->data, res->size) == -1) {
```
Comment (chaoyli) [inline be/src/env/env_hdfs.cpp:55]: What's the difference between hdfsPreadFully and hdfsRead?
Could all use the hdfsPreadFully instead of hdfsRead?

**Comment by chaoyli** [comment]
It's better to add a concise message about the optimization and the result.

File: be/src/env/env_hdfs.cpp (hunk 2 above)
Comment (imay) [reply be/src/env/env_hdfs.cpp:55]: Better to comment in this files to let people know why do it differently.

**Review by imay — DISMISSED** [review]
LGTM

File: be/src/env/env_hdfs.cpp (hunk 2 above)
Comment (miomiocat) [reply be/src/env/env_hdfs.cpp]: remove hdfsRead and keep hdfsPreadFully only after discussion offline. PTAL

**Comment by chaoyli** [comment]
run all

**Review by chaoyli — DISMISSED** [review]

**Review by chaoyli — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Pull request merged**